pgrx/src/datum/numeric_support/datum.rs

```diff
@@ -22,42 +22,48 @@ impl FromDatum for AnyNumeric {
         if is_null {
             None
         } else {
-            let numeric = pg_sys::pg_detoast_datum(datum.cast_mut_ptr()) as pg_sys::Numeric;
+            // Going back to Postgres v9.1, `pg_sys::NumericData` is really a "varlena" in disguise.
+            //
+            // We want to copy it out of the Postgres-allocated memory it's in and into something
+            // managed by Rust.
+            //
+            // First we'll detoast it and then copy the backing varlena bytes into a `Vec<u8>`.
+            // This is what we'll use later if we need to convert back into a postgres-allocated Datum
+            // or to provide a view over the bytes as a Datum.  The latter is what most of the AnyNumeric
+            // support functions use, via the `as_datum()` function.
+
+            // detoast
+            let numeric = pg_sys::pg_detoast_datum(datum.cast_mut_ptr());
             let is_copy = !std::ptr::eq(
-                numeric.cast::<pg_sys::NumericData>(),
-                datum.cast_mut_ptr::<pg_sys::NumericData>(),
+                numeric.cast::<pg_sys::varlena>(),
+                datum.cast_mut_ptr::<pg_sys::varlena>(),
             );
-            Some(AnyNumeric { inner: numeric, need_pfree: is_copy })
-        }
-    }
 
-    unsafe fn from_datum_in_memory_context(
-        mut memory_context: PgMemoryContexts,
-        datum: pg_sys::Datum,
-        is_null: bool,
-        _typoid: pg_sys::Oid,
-    ) -> Option<Self>
-    where
-        Self: Sized,
-    {
-        if is_null {
-            None
-        } else {
-            memory_context.switch_to(|_| {
-                // copy the Datum into this MemoryContext and then create the AnyNumeric over that
-                let copy = pg_sys::pg_detoast_datum_copy(datum.cast_mut_ptr());
-                Some(AnyNumeric { inner: copy.cast(), need_pfree: true })
-            })
+            // copy us into a Vec<u8>
+            let size = varsize_any(numeric);
+            let mut bytes = vec![0; size];
+            std::ptr::copy_nonoverlapping(numeric.cast::<u8>(), bytes.as_mut_ptr(), size);
```

Comment:
@thomcc how would this allocate a `Box<[u8]>` instead?

Reply:
Just convert to `Box<[u8]>` after -- It should be zero-cost in this case.